### PR TITLE
[PEPPER-1262] Enable export of null values for OncHistory

### PR DESCRIPTION
**Pull request by denniscunningham**
## Context
[PEPPER-1262]

The fundamental issue addressed with this PR is the inability of DSM to export a null `OncHistory.createdDate`. This is needed when  onc history is removed (currently only via direct DB updates, but probably a feature DSM should offer via the UI). It is also generally useful to be able to export null values in similar special cases, when identified.

This PR includes the following:

1. Change `ObjectTransformer` to allow export of null values, when specified.
2. Change `BaseMigrator` et. al. to enable `ObjectTransformer` specification.
3. Add a test case for exporting null  `OncHistory.createdDate` to `ElasticExportServiceTest`.
4. Update support for creating and updating `OncHistory` in `MedicalRecordTestUtil`.
5. Cleanup and restructure poorly organized code for creating `OncHistory`. This was overdue, and more restructuring would be beneficial, but for now just as much as make sense for a given test or feature. This also brings more untested code under test.
6. Create `OncHistoryService`. This was also overdue, and the service will likely grow in future implementations.
7. Improve Elastic export logging, including the updated ES document IDs (typically match DDP participant ID).

_Say a few words about the bigger picture beyond this PR. How do your changes improve life for our users? Link directly to the jira ticket (just type "DDP-..." and it'll automatically link) to help your teammates quickly get up to speed on the motivation for these changes._

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have added regression test labels to the jira ticket
- [ ] I have added verification steps to the jira ticket
- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

[PEPPER-1262]: https://broadworkbench.atlassian.net/browse/PEPPER-1262?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/OncHistory.java [hunk 1]
```diff
@@ -56,37 +53,35 @@ public OncHistory(long participantId, String created, String reviewed, String ch
...
-        String createdDate = SystemUtil.getDateFormatted(System.currentTimeMillis());
-        SimpleResult results = inTransaction((conn) -> {
-            SimpleResult dbVals = new SimpleResult();
+    /**
+     * Set oncHistory created date for participant iff created date is null
+     *
+     * @return true if the created date was set, false if created date was already set or oncHistory record
+     *              was not found
+     */
+    public static boolean setOncHistoryCreated(int participantId, String createdDate, String userId) {
```
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/OncHistory.java:62]: Is this and Dao changes to support test cases OR want to add a admin feature  ?
Maybe we dont need as admin feature because PEPPER-1262 is a one of edge case scenarios.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/ObjectTransformer.java [hunk 2]
```diff
@@ -8,22 +8,26 @@
...
 
-    public ObjectTransformer(String realm, BaseParser parser) {
-        this(realm);
-        this.parser = parser;
+    public ObjectTransformer(String realm) {
+        this(realm, false);
     }
 
-    public ObjectTransformer(String realm) {
+    public ObjectTransformer(String realm, boolean includeNulls) {
```
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/ObjectTransformer.java:28]: nice. we have a way to decide include or not include nulls

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/OncHistory.java (hunk 1 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/OncHistory.java:62]: This is to support test cases and to make the OncHistory code easier to work with in the future.

I think the actual feature we should support is allowing edit of onc history create date from the UI, but I can see how we may not want to offer that to end users. Perhaps only to pepper-admins? I will discuss with Whitney when she is back.

With this change if we make a OncHistory createDate change in DB, a subsequent export will work properly. As you mention, this is an edge case, so perhaps that is sufficient for now?

**Review by ssettipalli — APPROVED** [review]
looks ok to me
